### ISSUE TITLE
"Tree" and "Hourglass" algorithms for all-pairs shortest paths

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2197]
**What is the feature or improvement you would like to see?**

Implement the "Tree" or "Hourglass" algorithms for all-pairs shortest paths from the reference below. The paper appears to be well-written, and contains not only a theoretical analysis of the algorithm, but also an extensive empirical comparison. I think that there is a good chance that it works well in practice.

**Use cases for the feature**

I am looking for an algorithm that works well on dense graphs, i.e. the running time does not increase noticeably with the edge count when keeping the vertex count constant. The Floyd-Warshall algorithm ($O(V^3)$) has this property, but in practice it is slower than Dijkstra except for _very_ dense graphs. The reference claims an expected running time of $O(V^2 \log^2 V)$.

**References**

 - Andrej Brodnik, Marko Grgurovič, Rok Požar: Modifications of the Floyd-Warshall algorithm with nearly quadratic expected-time, https://doi.org/10.26493/1855-3974.2467.497

**Comment by Nivig1310** [comment]
Hello, I am interested in working on this but I am confused about what you would like implemented. Should I apply the Floyd-Warshall algorithm based on the tips given in the reference to optimize it?

**Comment by szhorvat** [comment]
Yes, the task is to implement at least one of the two algorithms described in the paper, ideally the faster one.  However, this is not a trivial task, so unless you are already comfortable with igraph, and experienced with graph algorithms, I suggest you start with a simpler issues first. I suggest #1842/#1843, or #1840 (which has stalled and is available). These would allow you to learn how to program with igraph without needing to also deal with a complex algorithm at the same time. It is up to you whether you start with something simple.

For this issue, you would need to:

 - Read the paper, understand the algorithms, summarize them briefly here.
 - Understand what is needed to create a production quality implementation that works on non-simple graphs
 - Implement one algorithm (or both). If you hourglass one, you would need to look into how to make it non-recursive, as recursion [risks stack-overflow](https://github.com/igraph/python-igraph/issues/563). A recursive version is fine for a starting point though.
 - Benchmark the implementation against Dijkstra and a plain Floyd-Warshall (#2194).

Here's what you need to know to get started: https://github.com/igraph/igraph/wiki/Quickstart-for-new-contributors

**Comment by szhorvat** [comment]
@rfulekjames Feel free to go ahead with this if you are interested. There is a benchmark for distance matrix calculations in `tests/benchmarks`. You can build benchmarks with `make build_benchmarks`, and you can run this one manually.  The final interface of the function is easily changed at the end, so I think we don't need to discuss that now (i.e. whether the method is selected with a parameter of `igraph_distance_floyd_warshall()` or it's a separate function). I should note now that I will have only very limited time in December though ...

**Comment by rfulekjames** [comment]
@szhorvat The "Tree Algorithm" from the paper gives best results for larger dense graphs.

 1 vcount=100, p=0.5, Unweighted, 200x                                            0.19s   0.19s      0s
|  2 vcount=100, p=0.5, Dijkstra, 200x                                              1.68s   1.66s   0.01s
|  3 vcount=100, p=0.5, Unweighted Floyd-Warshall, 200x                            0.573s  0.573s      0s
|  3 vcount=100, p=0.5, Unweighted Floyd-Warshall-tree-speedup, 200x               0.899s  0.899s      0s
|  4 vcount=100, p=0.5, Floyd-Warshall, 200x                                       0.705s  0.705s      0s
|  4 vcount=100, p=0.5, Floyd-Warshall-tree-speedup, 200x                            1.1s    1.1s      0s
|  5 vcount=100, p=0.5, Bellman-Ford (negative), 200x                               1.42s   1.42s      0s
|  6 vcount=100, p=0.5, Johnson (negative), 200x                                     1.8s    1.8s      0s
|  7 vcount=100, p=0.5, Floyd-Warshall (negative), 200x                            0.724s  0.724s      0s
|  7 vcount=100, p=0.5, Floyd-Warshall-tree-speedup (negative), 200x                 1.1s   1.09s   0.01s

|  1 vcount=500, p=0.1, Unweighted, 10x                                            0.254s  0.254s      0s
|  2 vcount=500, p=0.1, Dijkstra, 10x                                               2.49s   2.49s      0s
|  3 vcount=500, p=0.1, Unweighted Floyd-Warshall, 10x                              3.29s   3.29s      0s
|  3 vcount=500, p=0.1, Unweighted Floyd-Warshall-tree-speedup, 10x                 1.79s   1.79s      0s
|  4 vcount=500, p=0.1, Floyd-Warshall, 10x                                         3.59s   3.59s      0s
**|  4 vcount=500, p=0.1, Floyd-Warshall-tree-speedup, 10x                            2.04s   2.04s      0s**
|  5 vcount=500, p=0.1, Bellman-Ford (negative), 10x                                2.62s   2.62s      0s
|  6 vcount=500, p=0.1, Johnson (negative), 10x                                     2.52s   2.52s      0s
|  7 vcount=500, p=0.1, Floyd-Warshall (negative), 10x                              3.57s   3.57s      0s
**|  7 vcount=500, p=0.1, Floyd-Warshall-tree-speedup (negative), 10x                 2.03s   2.03s      0s**

|  1 vcount=1500, p=0.02, Unweighted, 1x                                           0.179s  0.179s      0s
|  2 vcount=1500, p=0.02, Dijkstra, 1x                                              1.78s   1.78s      0s
|  3 vcount=1500, p=0.02, Unweighted Floyd-Warshall, 1x                             8.56s   8.56s      0s
|  3 vcount=1500, p=0.02, Unweighted Floyd-Warshall-tree-speedup, 1x                3.11s   3.11s      0s
|  4 vcount=1500, p=0.02, Floyd-Warshall, 1x                                        8.85s   8.85s      0s
|  4 vcount=1500, p=0.02, Floyd-Warshall-tree-speedup, 1x                           3.52s   3.51s   0.01s
|  5 vcount=1500, p=0.02, Bellman-Ford (negative), 1x                               3.02s   3.02s      0s
|  6 vcount=1500, p=0.02, Johnson (negative), 1x                                    1.89s   1.89s      0s
|  7 vcount=1500, p=0.02, Floyd-Warshall (negative), 1x                             8.84s   8.84s      0s
|  7 vcount=1500, p=0.02, Floyd-Warshall-tree-speedup (negative), 1x                3.53s   3.53s      0s

**Comment by szhorvat** [comment]
That was very quick, and the results are exciting! Can you open a PR?

**Comment by rfulekjames** [comment]
The first "Tree"  algorithm was easy to implement. Only the main loop of the original Floyd-Warshall implementation had to be changed. Also using `igraph_tree_from_parent_vector()` function helped. 
The "hourglass" algorithm might be more intricate.

**Comment by gulshan-123** [comment]
Hi @szhorvat, while exploring similar issue #2822 (but not getting the idea), I saw "Hourglass" algorithm yet to be implemented. I was trying to code this for a few days, and gladly the code (initial draft) works.
Should I open a PR now, (as my 1 unrelated PR is pending #2817 )?
After the release of igraph 1.0, should it be targeted on develop or master?
